### ISSUE TITLE
Cannot upload scripts with CDATA

### ISSUE THREAD

**Issue by LeXofLeviafan** [JasonBarnabe/greasyfork#622]
Scripts using CDATA format inserts (e.g. Coffeescript) cannot be uploaded anymore because of syntax checker.

**Comment by JasonBarnabe** [comment]
Does any user script manager support Coffeescript?

**Comment by Martii** [comment]
@JasonBarnabe 

> Does any user script manager support Coffeescript?

You might also be looking for E4X XMLList compatibility too. At least TM 4.8.41 under `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chromium/75.0.3770.90 Chrome/75.0.3770.90 Safari/537.36` **currently does** with https://sourceforge.net/p/greasemonkey/wiki/Metadata_Block/#restructuring-with-e4x-xmllist . Still works here with a basic `alert` statement for the function *(which is a little surprising)*. Fx/SM *([Mozilla](https://developer.mozilla.org/docs/Archive/Web/E4X))* **currently does not**.

Ref:
* <a href="https://user-images.githubusercontent.com/114709/61657684-c3a9fe80-ac80-11e9-80cf-e5c0c273fd68.png"><img src="https://user-images.githubusercontent.com/114709/61657684-c3a9fe80-ac80-11e9-80cf-e5c0c273fd68.png" alt="E4X XMLList Unit Test screenshot" title="Click to enlarge...E4X XMLList Unit Test screenshot" width="101" height="54" /></a>

**Comment by LeXofLeviafan** [comment]
There is a CoffeeScript template in Tampermonkey (Settings tab, Userscripts section); which is how I found out about CDATA option. Incidentally, the language itself doesn't matter as its support is implemented via transpiling heredoc (CDATA) content using an external script (e.g. Babel).

**Comment by Martii** [comment]
> ... CoffeeScript template in Tampermonkey (Settings tab, Userscripts section) ...

Would probably be better to utilize a [Template Literal](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) for that rather than E4X... using the correct, industry standard, terminology. 

Unrelated but should be a concern none-the-less... MITM transpiling doesn't always yield decent results *(and sometimes unsafe especially since you mentioned "external")* but that's another story beyond the reach of this issue.

**Comment by LeXofLeviafan** [comment]
Multiline strings are ill-suited for describing actual code (having to escape every escaping backslash is a problem, for instance). More to the point, CDATA is, in fact, a valid way of representing heredoc code sources in a userscript, and therefore should not be blocked by a syntax checker.
Other concerns like whether or not someone considers runtime transpilation a good practice (which I believe is subjective at best, to begin with) is also not relevant to the issue in question.

**Comment by Martii** [comment]
So what I gather from all your responses, in summary you want GF, at the bare minimum, to support heredoc *(transpiling)* and all the additional security filtering the site does for multiple languages *(inclusive of supersets)* instead of vanilla Userscripts written in pure JavaScript.

The core of [the initial citation](https://github.com/JasonBarnabe/greasyfork/issues/622#issuecomment-513914400) still remains as defacto. Some browsers support E4X some do not *(unclear atm whether this is a shim in TM or not too)*. It can be a great tool in the kit for fingerprinting a particular set of users under a very finite system. Always relevant when it comes to Privacy and Security... and of course Copyright. All completely relevant when expanding a Userscript to use other languages. Btw we utilizes some of the components you have mentioned so we are all aware of the implications and consequences but not usually external. You may draw that conclusion rather then taking guesses at any intentions. :)

Hopefully everything here will give Jason enough information to make an informed decision to support hosting a description of a Userscript or not. I'll be eagerly awaiting some other input from him to read his views for support.

**Comment by LeXofLeviafan** [comment]
An opensource script distributed via CDN is unlikely to be a security problem (and those paranoid enough to avoid them would do so without external prompting); same thing can be said about extensions (such as userscript managers). I don't think it's any more a liability than, say, website cookies.

>to make an informed decision to support hosting a description of a Userscript or not

It was supported just a couple weeks ago; the reason I opened this issue is because it prevents me from updating an existing userscript. And I don't believe it's a good thing to just break what's working; enforced syntax check with no way of circumventing it should only be done when it doesn't prevent any actually valid input.

**Comment by Martii** [comment]
> An opensource script distributed via CDN is unlikely to be a security problem...

"Hehe" that's all I have to say. ;)

> It was supported just a couple weeks ago

Already entered this assumption based off your initial report. I can already deduce that it used to be supported but now is not. So the the terminology is to support or not to support. I can say historically E4X has been problematic with security which is one of the many reasons it was deprecated for more modern technologies. Things change and sometimes back. Depends on who is behind the wheel. I'll be watching. :smile_cat:

**Comment by JasonBarnabe** [comment]
@LeXofLeviafan Can you provide an example of a script that no longer works?

Honestly, I'm leaning toward not supporting this just for compatibility reasons, but I did not intentionally remove support with syntax checker. I had just never heard of this.

**Comment by LeXofLeviafan** [comment]
Here's the example taken directly from Tampermonkey settings screen (with boilerplate removed and script URL replaced with one from a CDN)
```
// ==UserScript==
// @require      https://cdn.jsdelivr.net/npm/coffeescript@2.4.1/lib/coffeescript-browser-compiler-legacy/coffeescript.js
// ==/UserScript==

var inline_src = (<><![CDATA[

    console.log 42    # Your code here

]]></>).toString();
var compiled = this.CoffeeScript.compile(inline_src);
eval(compiled);
```
And if you want a real world script, here's mine (the one I was trying to update):
https://greasyfork.org/ru/scripts/386681-backloggery-interop/code

**Comment by JasonBarnabe** [comment]
OK, so Coffeescript is unrelated. The issue is that you're using E4X stuff and the syntax checker (which uses V8's engine) doesn't like it.

Considering E4X is effectively dead, and there are good alternatives (other JS string syntaxes, transpile it offline and just post your JS), then I'm going to say this is not going to be supported.

**Comment by LeXofLeviafan** [comment]
> good alternatives
Correction: _largely inferior_ alternatives.
